toolchain/check/call.cpp

```diff
@@ -37,6 +41,101 @@ enum class EntityKind : uint8_t {
 };
 }  // namespace
 
+// Performs overloading resolution for a call to an overloaded C++ set. A set
+// with a single non-templated function is still considered to be an overload
+// set, and goes through the same rules for checking the viability of the
+// function. Uses Clang to find the best viable function for the call. Returns
+// the resolved function, or `nullopt` if overload resolution failed.
+static auto PerformOverloadResolution(Context& context, SemIR::LocId loc_id,
+                                      SemIR::InstId callee_id,
+                                      llvm::ArrayRef<SemIR::InstId> arg_ids)
+    -> std::optional<SemIR::InstId> {
+  Diagnostics::AnnotationScope annotate_diagnostics(
+      &context.emitter(), [&](auto& builder) {
+        CARBON_DIAGNOSTIC(InCallToOverloadedCppFunction, Note,
+                          "in call to a Cpp (overloaded) function here");
+        builder.Note(loc_id, InCallToOverloadedCppFunction);
+      });
+
+  // Map Carbon call argument types to C++ types.
+  llvm::SmallVector<clang::Expr*, 12> arg_exprs;
+  for (SemIR::InstId arg_id : arg_ids) {
+    auto arg_cpp_type = MapToCppType(context, arg_id);
+    if (!arg_cpp_type) {
+      CARBON_DIAGNOSTIC(CallArgTypeNotSupported, Error,
+                        "call arg of type {0} is not supported", TypeOfInstId);
+      context.emitter().Emit(loc_id, CallArgTypeNotSupported, arg_id);
+      return std::nullopt;
+    }
+    auto* arg_expr = new (context.ast_context()) clang::OpaqueValueExpr(
+        clang::SourceLocation(), arg_cpp_type->getNonReferenceType(),
+        clang::ExprValueKind::VK_LValue);
+    arg_exprs.emplace_back(arg_expr);
+  }
+
+  auto fn_type_inst =
+      context.types().GetAsInst(context.insts().Get(callee_id).type_id());
+  auto fn_type = fn_type_inst.TryAs<SemIR::OverloadedCppFunctionType>();
+  if (!fn_type) {
+    return std::nullopt;
+  }
+  auto overloaded_fn =
```

Comment:
Can this be:

```suggestion
  const auto& overloaded_fn =
```

Or, even better:

```suggestion
  const OverloadedCppFunction& overloaded_fn =
```

Reply:
Done.